Software-Construction/Factory-And-Abstract-Factory/Graphics/Screen.java

```diff
@@ -0,0 +1,177 @@
+package Graphics;
+
+import java.util.ArrayList;
+import java.util.Date;
+import java.util.List;
+
+import Graphics.Shapes.ShapeType;
+
+/**
+ * @author Niraj Gupta
+ * Class to represent the screen on which the shapes are drawn.
+ */
+public class Screen{
+	private List<Shapes> listOfShapes = new ArrayList<>();
+	private final double XMAX, YMAX;
+
+	/**
+	 * Parameterized Constructor
+	 * @param xMax : The maximum length of the screen.
+	 * @param yMax : The maximum breadth of the screen.
+	 */
+	public Screen(double xMax, double yMax){
+		XMAX = xMax;
+		YMAX = yMax;
+	}
+
+	/**
+	 * Method to add shape on the screen
+	 * @param shape : The shape which is to be added on the screen.
+	 */
+	public void addShape(Shapes shape){
+		if(shape == null){
+			throw new AssertionError("Null object");
+		}
+		Point origin = shape.getOrigin();
+		if(origin.getX() < 0 || origin.getX() > XMAX || origin.getY() < 0 || origin.getY() > YMAX){
+			throw new AssertionError("Cannot create shape outside the screen");
+		}
+		listOfShapes.add(shape);
+		shape.setTimestamp(new Date());
+	}
+
+	/**
+	 * Method to remove the shape from the screen.
+	 * @param shape : The shape which is to be removed from the screen.
+	 */
+	public void deleteShape(Shapes shape){
+		if(listOfShapes.contains(shape)){
+			listOfShapes.remove(shape);
+		}else{
+			throw new AssertionError("The shape is not present on screen!");
+		}
+	}
+
+	/**
+	 * Method to remove all the shapes from the screen belonging to a particular screen type.
+	 * @param shapeType : The type of the shape which is to be removed.
+	 */
+	public void deleteShapeType(ShapeType shapeType){
+		boolean flag = false;
+		List<Shapes> removeList = new ArrayList<>();
+
+		for(Shapes shape : listOfShapes){
+			if(shape.getShapeType() == shapeType){
+				flag = true;
+				removeList.add(shape);
+			}
+		}
+		listOfShapes.removeAll(removeList);
+		if(!flag){
+			throw new AssertionError("No object of the given type!");
+		}
+	}
+
+	/**
+	 * Method to sort the shapes on the basis of their area.
+	 * @return : The sorted list of shape.
+	 */
+	public List<Shapes> sortByArea(){
+		List<Shapes> sortedListByArea = new ArrayList<>(listOfShapes);
+		int noOfObjects = listOfShapes.size();
+
+		if(noOfObjects == 0){
+			throw new AssertionError("List is empty! Cannot sort");
+		}
+
+		for(int i = 0; i < noOfObjects; i++){
+			for(int j = 0; j < noOfObjects - 1 - i; j++){
+				if(sortedListByArea.get(j).getArea() > sortedListByArea.get(j + 1).getArea()){
+					Shapes temp = sortedListByArea.get(j);
+					sortedListByArea.set(j, sortedListByArea.get(j + 1));
+					sortedListByArea.set(j + 1, temp);
+				}
+			}
+		}
+		return sortedListByArea;
+	}
+
+	/**
+	 * Method to sort the shapes on the basis of their perimeter.
+	 * @return : The sorted list of the shape.
+	 */
+	public List<Shapes> sortByPerimeter(){

```

Comment:
same as above method it could have been breaked into submethods, doing this would have also made you able to keep the common code in one place and reduce code complexity as well as no of code lines

Reply:
I should have written a different swap method. My fault.